doc/hare-dtm-hld.md

```diff
@@ -0,0 +1,614 @@
+with the help of maintaining the process states in Consul and respective
+callbacks for DTM state updates. When Motr process completes DTM, it
+sends M0_CONF_HA_PROCESS_STARTED notification to Hare. On this event,
+Hare then broadcasts ONLINE to all the Motr processes via BQ. In case of any
+failure during the recovery process Hare handles the situation again with the
+help of states saved in the Consul. This is also described in the `DTM
+recovery states section`.
+
+### Concurrent failures
+Following concurrent failures can be considered,
+#### Recovering process failure
+In this case Hare receives a failure notification for the recovering process.
+If Hare has already update the process's ha state to M0_NC_RECOVERING and its
+status to M0_CONF_HA_PROCESS_STARTING, then on the next restart of the failed
+Motr process Hare refers to the process's saved states to take a decision to
+re-trigger DTM recovery. This is also illustrated in `DTM Recovery States`
+section.
+If Hare finds the saved ha states as M0_NC_FAILED or M0_NC_TRANSIENT and
+process status as M0_CONF_HA_PROCESS_STOPPED, then Hare broadcasts
+M0_NC_TRANSIENT for that Motr process and restarts DTM recovery process.
+<!--
+@startuml
+autonumber
+ha_consul -> hax1_RC: notify Motr2 failure
+ha_consul -> hax2: notify Motr2 failure
+ha_consul -> hax3: notify Motr2 failure
+hax1_RC -> consul: post a EQ event to update failvec and process ha state
+hax2 -> consul: post a EQ event to update failvec and process ha state
+hax3 -> consul: post a EQ event to update failvec and process ha state
+hax1_RC -> consul: process EQ, update failvec and Motr2 ha state, discards duplicates
+Motr2 -> hax2: restarts, requests entrypoint
+hax2 -> consul: fetches failvec for current Motr2 epoch
+hax2 -> Motr2: send entrypoint reply with ha states
+hax2 -> consul: update Motr2 ha state to M0_NC_TRANSIENT and broadcast (BQ) the same
+Motr2 -> hax2: notifies M0_CONF_HA_PROCESS_STARTING
+hax2 -> consul: updates Motr2 process status to M0_CONF_HA_PROCESS_STARTING and ha state to M0_NC_RECOVERING
+hax2 -> hax2: wait for dtm recovery to complete
+destroy Motr2
+ha_consul -> hax1_RC: notify Motr2 failure
+ha_consul -> hax2: notify Motr2 failure
+ha_consul -> hax3: notify Motr2 failure
+hax1 -> consul: get Motr2 ha state
+hax2 -> consul: get Motr2 ha state
+hax3 -> consul: get Motr2 ha state
+    loop if Motr2 ha state is M0_NC_RECOVERING
+        hax1 -> hax1: do nothing, let Motr2 local, hax2 handle
+        hax2 -> hax2: wait for Motr2 to restart
+        hax3 -> hax3: do nothing, let Motr2 local hax2 handle
+    end
+Motr2 -> hax2: restarts, requests entrypoint
+hax2 -> Motr2: sends entrypoint reply, ha states
+hax2 -> consul: post BQ event to broadcast M0_NC_TRANSIENT
+hax2 -> consul: post BQ event to broadcast M0_NC_RECOVERING
+hax2 -> hax2: wait for dtm recovery to complete
+@enduml
+-->
+![img](recovering-process-failure.png)
+
+#### Participating process failure
```

Comment:
How Hare will understand which m0d processes are "participating"?

In other words, how can we know which m0d processes will participate in the same DTM recovery operation?

Reply:
I think that is mentioned in `DTM participants` section of functional specifications.

Reply:
Updated documentation and sequence diagram.